src/StringTools/OpportunisticIntern.cs

```diff
@@ -0,0 +1,295 @@
+            /// seen. The higher the number the better the payoff if the string had been hardcoded.
+            /// </summary>
+            private Dictionary<string, int> _missedHardcodedStrings;
+
+#endregion
+
+            public WeakStringCacheInterner(bool gatherStatistics)
+            {
+                if (gatherStatistics)
+                {
+                    _missedHardcodedStrings = new Dictionary<string, int>();
+                }
+                _gatherStatistics = gatherStatistics;
+            }
+
+            /// <summary>
+            /// Intern the given internable.
+            /// </summary>
+            public string InterningToString(ref InternableString candidate)
+            {
+                if (_gatherStatistics)
+                {
+                    return InternWithStatistics(ref candidate);
+                }
+                else
+                {
+                    TryIntern(ref candidate, out string result);
+                    return result;
+                }
+            }
+
+            /// <summary>
+            /// Returns a string with human-readable statistics.
+            /// </summary>
+            public string FormatStatistics()
+            {
+                StringBuilder result = new StringBuilder(1024);
+
+                string title = "Opportunistic Intern";
+
+                if (_gatherStatistics)
+                {
+                    result.AppendLine(string.Format("\n{0}{1}{0}", new string('=', 41 - (title.Length / 2)), title));
+                    result.AppendLine(string.Format("||{0,50}|{1,20:N0}|{2,8}|", "Hardcoded Hits", _hardcodedInternHits, "hits"));
+                    result.AppendLine(string.Format("||{0,50}|{1,20:N0}|{2,8}|", "Hardcoded Rejects", _rejectedStrings, "rejects"));
+                    result.AppendLine(string.Format("||{0,50}|{1,20:N0}|{2,8}|", "WeakStringCache Hits", _regularInternHits, "hits"));
+                    result.AppendLine(string.Format("||{0,50}|{1,20:N0}|{2,8}|", "WeakStringCache Misses", _regularInternMisses, "misses"));
+                    result.AppendLine(string.Format("||{0,50}|{1,20:N0}|{2,8}|", "Eliminated Strings*", _internEliminatedStrings, "strings"));
+                    result.AppendLine(string.Format("||{0,50}|{1,20:N0}|{2,8}|", "Eliminated Chars", _internEliminatedChars, "chars"));
+                    result.AppendLine(string.Format("||{0,50}|{1,20:N0}|{2,8}|", "Estimated Eliminated Bytes", _internEliminatedChars * 2, "bytes"));
+                    result.AppendLine("Elimination assumes that strings provided were unique objects.");
+                    result.AppendLine("|---------------------------------------------------------------------------------|");
+
+                    IEnumerable<string> topMissingHardcodedString =
+                        _missedHardcodedStrings
+                        .OrderByDescending(kv => kv.Value * kv.Key.Length)
+                        .Take(15)
```

Comment:
What if there are under 15?

And where did 15 come from?

And should the Where come before the Take?

Reply:
> What if there are under 15?

Then everything will be returned. From `Enumerable.Take<>` docs:
_If count exceeds the number of elements in source, all elements of source are returned._

> And where did 15 come from?

Just an arbitrary number to make the output reasonable in size. I copied this from the old code.

> And should the Where come before the Take?

Yes, it makes more sense that way. Fixed!